File: src/main/java/info/skyblond/archivedag/repo/CertRepository.java
package info.skyblond.archivedag.repo;

import info.skyblond.archivedag.model.entity.CertEntity;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;

import java.sql.Timestamp;
import java.util.List;

public interface CertRepository extends JpaRepository<CertEntity, String> {
    boolean existsBySerialNumber(String serialNumber);

    Page<CertEntity> findAllByUsernameContainingAndIssuedTimeBetweenAndExpiredTimeBetween(
            String ownerKeyword, Timestamp issueStart, Timestamp issueEnd,
            Timestamp expireStart, Timestamp expireEnd, Pageable pageable);

    Page<CertEntity> findAllByUsernameAndIssuedTimeBetweenAndExpiredTimeBetween(
            String owner, Timestamp issueStart, Timestamp issueEnd,
            Timestamp expireStart, Timestamp expireEnd, Pageable pageable);

    boolean existsBySerialNumberAndUsername(String serialNumber, String username);

    CertEntity findBySerialNumber(String serialNumber);

    @Modifying
    @Query("update CertEntity c set c.status = ?2 where c.serialNumber = ?1")
    void updateCertStatus(String serialNumber, CertEntity.Status status);

    @Modifying
    void deleteBySerialNumber(String serialNumber);

    List<CertEntity> findAllByUsername(String username);

    Page<CertEntity> findAllByUsername(String username, Pageable pageable);

}

File: backend/src/main/kotlin/info/skyblond/archivedag/arstue/repo/CertRepository.kt
package info.skyblond.archivedag.arstue.repo

import info.skyblond.archivedag.arstue.entity.CertEntity
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository

@Repository
interface CertRepository : JpaRepository<CertEntity, String> {
    fun existsBySerialNumber(serialNumber: String): Boolean

    fun findAllByUsernameContaining(ownerKeyword: String, pageable: Pageable): Page<CertEntity>

    fun findAllByUsername(owner: String, pageable: Pageable): Page<CertEntity>

    fun existsBySerialNumberAndUsername(serialNumber: String, username: String): Boolean

    fun findBySerialNumber(serialNumber: String): CertEntity?

    @Modifying
    @Query("update CertEntity c set c.status = ?2 where c.serialNumber = ?1")
    fun updateCertStatus(serialNumber: String, status: CertEntity.Status)

    @Modifying
    fun deleteBySerialNumber(serialNumber: String)

    @Modifying
    fun deleteAllByUsername(username: String)

}
